From 1dec452c446aaa8ab37c75bf0bcae9f8ea8b3fea Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Fri, 13 Sep 2024 22:39:10 +0900
Subject: [PATCH 04/19] V4L2Buffer: Allocate correct v4l2 buffers for queues

For single plane queue, need to fill v4l2_planes_[0] with
correct size quried from v4l2 driver.

Upstream-Status: Inappropriate [NXP specific]
---
 media/gpu/v4l2/v4l2_queue.cc | 39 +++++++++++++++++++++++++-----------
 1 file changed, 27 insertions(+), 12 deletions(-)

diff --git a/media/gpu/v4l2/v4l2_queue.cc b/media/gpu/v4l2/v4l2_queue.cc
index 4851e5b5255c0..8bcf8101f3a20 100644
--- a/media/gpu/v4l2/v4l2_queue.cc
+++ b/media/gpu/v4l2/v4l2_queue.cc
@@ -237,24 +237,28 @@ V4L2Buffer::V4L2Buffer(const IoctlAsCallback& ioctl_cb,
   DCHECK(V4L2_TYPE_IS_MULTIPLANAR(type));
   DCHECK_LE(format.fmt.pix_mp.num_planes, std::size(v4l2_planes_));
 
-  memset(&v4l2_buffer_, 0, sizeof(v4l2_buffer_));
-  memset(v4l2_planes_, 0, sizeof(v4l2_planes_));
-  v4l2_buffer_.m.planes = v4l2_planes_;
-  // Just in case we got more planes than we want.
-  v4l2_buffer_.length =
-      std::min(static_cast<size_t>(format.fmt.pix_mp.num_planes),
-               std::size(v4l2_planes_));
+  if (V4L2_TYPE_IS_MULTIPLANAR(type)) {
+    memset(&v4l2_buffer_, 0, sizeof(v4l2_buffer_));
+    memset(v4l2_planes_, 0, sizeof(v4l2_planes_));
+    v4l2_buffer_.m.planes = v4l2_planes_;
+    // Just in case we got more planes than we want.
+    v4l2_buffer_.length =
+        std::min(static_cast<size_t>(format.fmt.pix_mp.num_planes),
+                std::size(v4l2_planes_));
+  }
   v4l2_buffer_.index = buffer_id;
   v4l2_buffer_.type = type;
   v4l2_buffer_.memory = memory;
-  plane_mappings_.resize(v4l2_buffer_.length);
+  plane_mappings_.resize(V4L2_TYPE_IS_MULTIPLANAR(type) ? v4l2_buffer_.length : 1);
 }
 
 V4L2Buffer::~V4L2Buffer() {
   if (v4l2_buffer_.memory == V4L2_MEMORY_MMAP) {
     for (size_t i = 0; i < plane_mappings_.size(); i++) {
       if (plane_mappings_[i] != nullptr) {
-        munmap(plane_mappings_[i], v4l2_buffer_.m.planes[i].length);
+        unsigned int length = V4L2_TYPE_IS_MULTIPLANAR(v4l2_buffer_.type) ?
+            v4l2_buffer_.m.planes[i].length : v4l2_buffer_.length;
+        munmap(plane_mappings_[i], length);
       }
     }
   }
@@ -268,6 +272,13 @@ bool V4L2Buffer::Query() {
     return false;
   }
 
+  if (!V4L2_TYPE_IS_MULTIPLANAR(v4l2_buffer_.type)) {
+    v4l2_planes_[0].bytesused = v4l2_buffer_.bytesused;
+    v4l2_planes_[0].length = v4l2_buffer_.length;
+    v4l2_planes_[0].data_offset = 0;
+    memcpy (&v4l2_planes_[0].m, &v4l2_buffer_.m, sizeof (v4l2_buffer_.m));
+  }
+
   DCHECK(plane_mappings_.size() == v4l2_buffer_.length);
 
   return true;
@@ -291,9 +302,13 @@ void* V4L2Buffer::GetPlaneMapping(const size_t plane) {
     return nullptr;
   }
 
-  p = mmap_cb_.Run(nullptr, v4l2_buffer_.m.planes[plane].length,
+  unsigned int length = V4L2_TYPE_IS_MULTIPLANAR(v4l2_buffer_.type) ?
+      v4l2_buffer_.m.planes[plane].length : v4l2_planes_[plane].length;
+  unsigned int mem_offset = V4L2_TYPE_IS_MULTIPLANAR(v4l2_buffer_.type) ?
+      v4l2_buffer_.m.planes[plane].m.mem_offset : v4l2_planes_[plane].m.mem_offset;
+  p = mmap_cb_.Run(nullptr, length,
                    PROT_READ | PROT_WRITE, MAP_SHARED,
-                   v4l2_buffer_.m.planes[plane].m.mem_offset);
+                   mem_offset);
   if (p == MAP_FAILED) {
     VPLOGF(1) << "mmap() failed: ";
     return nullptr;
@@ -1177,7 +1192,7 @@ size_t V4L2Queue::AllocateBuffers(size_t count,
     VQLOGF(1) << "Cannot get format.";
     return 0;
   }
-  planes_count_ = format->fmt.pix_mp.num_planes;
+  planes_count_ = V4L2_TYPE_IS_MULTIPLANAR(format->type) ? format->fmt.pix_mp.num_planes : 1;
   DCHECK_LE(planes_count_, static_cast<size_t>(VIDEO_MAX_PLANES));
 
   __u8 flags = incoherent ? V4L2_MEMORY_FLAG_NON_COHERENT : 0;
-- 
2.34.1

